spack.ci: Refactor process_command for Cross Platform support

johnwparent opened the pull request:
Current shell scripts generated by `process_command` are not compatible with Windows. This PR refactors `process_command` for cross platform support.
On Linux-like platforms, this should effectively be a no-op, on Windows, Spack now generates powershell scripts. Powershell was selected because this is the only shell we can run Gitlab-CI jobs on.
We utilize the `Set-PSDebug` commandlet to replicate `set -x` as it's the closest direct analog. A combination of `ErrorActionPreference` and manual checks for `$LastExitCode` are used to replicate the behavior of `set -e` in Powershell. 
This is done because powershell has no overt mechanism to exit a script on arbitrary command failure. `ErrorActionPreference` handles Powershell commandlets, but has no handling for native commands (exes). We handle this by checking the exit code of exes via `LastExitCode`. 

This PR also updates `reproduce_ci_job` to handle the changes in `process_command` as well as updates the method to better support running on Powershell.

Refactored the whole module to use f-strings. I refactored about 10 of them in the two methods I was working on, and realized there were only a few more in the rest of the module so I figured why not.


For context: This PR previously attempted to convert the above to Python instead of platform specific shells. This approach was deemed overly fragile.

haampie commented:
What happened to the idea of getting rid of this shell script?

kwryankrattiger committed:
CI: Disable downloading artifacts from upstream jobs

kwryankrattiger commented:
The code here makes sense to me. I am going to merge this into #42657 and make sure it runs as expected.

kwryankrattiger commented on lib/spack/spack/cmd/ci.py:
```diff
@@ -580,14 +585,14 @@ def ci_rebuild(args):
         # apparently there's a race when spack bootstraps? do it up front once
-        [SPACK_COMMAND, "-e", env.path, "bootstrap", "now"],
+        [SPACK_COMMAND, "-e", encode_path(env.path), "bootstrap", "now"],
         [
```
We can just remove all of the makefile parts here.

johnwparent replied:
Works for me!

scheibelp replied:
It looks like that hasn't happened yet (still see makefile references below). This could be replaced by a `spack install`, although I don't know why it was done with `depfile` here before so there could be some performance differences

johnwparent replied:
I haven't taken out the makefile stuff yet as I'm waiting to see if #39529 would progress (Harmen wrote this initially, and that PR of his also removes this, given that I think he's the best person to take this out), worst case I cherry-pick the relevant parts of that PR over to this one.

johnwparent replied:
The `depfile` was used to improve parallel installation of dependencies, but as I understand it, that did not end up being the location of the performance bottleneck the `depfile` was intended to improve, so we can refactor away from  `depfile` approach.

scheibelp replied:
What confused me in this PR is that this includes changes to a function that still cannot run (on Windows) given the changes in the PR: IMO the changes in this file should move to another PR.

scheibelp commented on lib/spack/spack/ci.py:
```diff
@@ -2020,25 +2020,51 @@ def process_command(name, commands, repro_dir, run=True, exit_on_failure=True):
+    args_to_string = lambda args: f'retcode = os.system("{" ".join(args)}")\n{fail_on_bad_return}'
+    full_command = "\n".join(map(args_to_string, commands))
+    # unfortunately Windows cannot arbitrarily directly execute python files
+    # so we wrap it in a mixed language batch file
```
I don't think this is true anymore (i.e. that this is a mixed batch/python language script): the extension is always `.py` and it looks like the expectation is that users run this with `python ./the-script.py` instead of `./the-script.py` (on both Linux and Windows).

johnwparent replied:
Yes, missed this docstring on refactor, should be taken out.

scheibelp commented on lib/spack/spack/ci.py:
```diff
@@ -2055,15 +2081,19 @@ def process_command(name, commands, repro_dir, run=True, exit_on_failure=True):
-            cmd_process = subprocess.Popen(["/bin/sh", "./{0}".format(script)])
+            # on Linux sh can drive the script thanks to the shebang
+            # On Windows we have no shebang, so we rely on "python"
+            # as an executor
```
This comment is worded WRT how things were before this PR which makes it confusing - the following would be more clear: "we *could* use sh on Linux but can't on Windows, so instead we use python everywhere".

johnwparent replied:
Makes sense, will refactor.

scheibelp commented on lib/spack/spack/ci.py:
```diff
@@ -2020,25 +2020,51 @@ def process_command(name, commands, repro_dir, run=True, exit_on_failure=True):
 
     Returns: the exit code from processing the command
     """
```
`reproduce_ci_job` creates a set of commands called `entrypoint_script` which includes

```
. share/spack/setup-env.sh
```

I'm confused why that doesn't have to change here.

johnwparent replied:
That may need to change in general, but this PR was more targeted toward changing the `process_command` workflow, rather than the callers who use it. Ryan and I will likely refactor this in the larger CI support PR.

scheibelp replied:
I suppose then that whatever processes are using this PR right now, they don't happen to invoke `reproduce_ci_job`? I can see the value of updating the minimum set, but I think the functions which are not essential should have a `TODO` comment on them.

scheibelp replied:
Update: it has been pointed out to me that the goal of this PR is to refactor the underlying script invocation mechanism: it does not intend to make the CI commands runnable on Windows.

scheibelp reviewed (changes requested):
~It looks like this is incomplete, so I'm leaving some higher-level comments for now.~ (this higher level concern will be resolved by https://github.com/spack/spack/pull/39739#discussion_r1500022930)

scheibelp commented on lib/spack/spack/ci.py:
```diff
@@ -2020,25 +2020,51 @@ def process_command(name, commands, repro_dir, run=True, exit_on_failure=True):
+        # Suppress all exceptions
+        pass
+    else:
+        sys.__excepthook__(exctype, value, traceback)
```
A few things:

* What doesn't fall under BaseException that you want to let through?
* I think if this strategy persists,
  * you should store a reference to the original `sys.excepthook` and use that vs. calling `sys.__excepthook__`
  * Add a comment about what this is doing e.g. `# suppress all exceptions`
* I would think that generally using `os.system` to run commands shouldn't throw exceptions unless the commands are malformed, which is the sort of thing we'd want to correct, so overall I'm not clear on what cases this is supposed to handle

johnwparent replied:
Looping context from Slack thread:
This was an artifact from a previous iteration, it's not longer necessary and is being removed.

johnwparent commented:
@scheibelp I believe I have addressed all outstanding comments. Ready for another round.

I'm going to pull the refactor of the test modules and `cmd/ci` module into separate PRs as discussed.

scheibelp commented on lib/spack/spack/test/cmd/ci.py:
```diff
@@ -2023,7 +2023,7 @@ def fake_download_and_extract_artifacts(url, work_dir):
-    assert os.path.exists(os.path.join(os.path.realpath(working_dir.strpath), "start.sh"))
+    assert os.path.exists(os.path.join(os.path.realpath(working_dir.strpath), "start.py"))
     # Make sure we tell the suer where it is when not in interactive mode
     expect_out = "$ {0}/start.sh".format(os.path.realpath(working_dir.strpath))
```
The instructions generated by `reproduce_ci_job` should change and this test should also change (I think `expect_out` should be `$ python start.py`). I also see a reference to `install.sh` in that method.

johnwparent replied:
@scheibelp thanks for pointing this out, on investigation I think executing https://github.com/spack/spack/pull/39739/files#diff-49f4a785e6a8295a5cd5af6a072cde881c00aecefe40fadf40abb33d5c35f6f5L1947 with the current approach in this PR will result in incorrect/undesirable behavior. We expect to pass the docker entrypoint the argument "bash" which I assume is meant to handle the interactive session in the reproducer from the `exec $@` command. Passing `bash` (or any argument) to the entrypoint.py script will have no effect and `os.system` will not allow for an interactive session via `exec`. 

This will require a bit more thought/more extensive refactor than I think either of us had initially planned on. 
Currently working on the best way to approach this. The test has been refactored and references to `.sh` scripts removed in favor of `.py` scripts, just working out how best to replicate expected behavior.

scheibelp reviewed (changes requested):
There's an inconsistency with the instructions generated by some methods now that `process_command` generates `.py` files instead of `.sh` scripts.

johnwparent committed:
ci: refactor process_command

Refactor process command to generate platform specific shells

Use sh on linux like platforms
Use powershell on Windows

Update ci.reproduce_ci_job to properly use new process_command
Update reproduce_ci_job for Windows

johnwparent commented:
@kwryankrattiger See new refactor (went ahead and did reproduce anyway)

scheibelp commented on lib/spack/spack/ci.py:
```diff
@@ -2020,50 +2035,66 @@ def process_command(name, commands, repro_dir, run=True, exit_on_failure=True):
             exit_code = cmd_process.returncode
         except (ValueError, subprocess.CalledProcessError, OSError) as err:
-            tty.error("Encountered error running {0} script".format(name))
+            tty.error(f"Encountered error running {name} script")
```
(no change needed) Note: every change below this in this file is a cosmetic refactor of `x.format()` into `f"x"`

johnwparent replied:
Thank you for doing this, I should have included it in the PR description.

scheibelp commented on lib/spack/spack/ci.py:
```diff
@@ -2020,50 +2035,66 @@ def process_command(name, commands, repro_dir, run=True, exit_on_failure=True):
-    os.chmod(copy_path, st.st_mode | stat.S_IEXEC)
+    if not IS_WINDOWS:
+        st = os.stat(copy_path)
+        os.chmod(copy_path, st.st_mode | stat.S_IEXEC)
```
(question) Since the script is being passed to an interpreter, does the `/bin/sh` version require this?

johnwparent replied:
I think the permissions change is required for `copy_path` (since it's the script that persists and might be used elsewhere by another actor) but is not for `script`, since, as you point out, it's being driven by `/bin/sh` anyway.

scheibelp commented on lib/spack/spack/ci.py:
```diff
@@ -1734,7 +1731,7 @@ def setup_spack_repro_version(repro_dir, checkout_commit, merge_commit=None):
 
             if git.returncode != 0:
-                tty.error("Unable to merge {0}".format(merge_commit))
+                tty.error(f"Unable to merge {merge_commit}")
```
(no change needed) Note: every change above here is a refactoring of `x.format()` to `f"x"`, minus the global definition of `IS_WINDOWS`.

scheibelp commented on lib/spack/spack/ci.py:
```diff
@@ -2020,50 +2035,66 @@ def process_command(name, commands, repro_dir, run=True, exit_on_failure=True):
+"""
+        if IS_WINDOWS
+        else ""
+    )
```
(question) If we're doing this after each command in a powershell script, is `$ErrorActionPreference = "Stop"` redundant? I would only be right about this if failed PWSH commandlets have a nonzero exit code.

johnwparent replied:
Powershell commandlets have "non-terminating errors" (i.e. if you manually write to stderr w\ `Write-Error` during a script run). Some will terminate in this way even when they've failed, but their exit code is still 0. Both cases are required.

scheibelp commented on lib/spack/spack/ci.py:
```diff
@@ -2020,50 +2035,66 @@ def process_command(name, commands, repro_dir, run=True, exit_on_failure=True):
+    )
+    # Create a string [command 1] \n [command 2] \n ... \n [command n] with
+    # commands composed into a platform dependent shell script, pwsh on Windows,
+    args_to_string = lambda args: f'{" ".join(args)}\n{catch_exe_failure.format(" ".join(args))}'
```
Can you refactor this into something like

```
def invoke_next_cmd_and_check_for_error():
    result = [
        " ".join(args)
    ]
    if catch_exe_failure:
        result.append(catch_exe_failure)
```

johnwparent replied:
Yeah this is messy, on it.

scheibelp commented on lib/spack/spack/ci.py:
```diff
@@ -2020,50 +2035,66 @@ def process_command(name, commands, repro_dir, run=True, exit_on_failure=True):
+    full_command = "\n".join(map(args_to_string, commands))
+    # Write the command to a python script
+    script = f"{name}.{ext}"
     with open(script, "w") as fd:
```
Since every line here is abstracted, IMO it would be more clear to split on `IS_WINDOWS` here like:

```
if IS_WINDOWS:
    script = [
        ...
    ]
else:
    script = [
        ...
    ]
```

johnwparent replied:
Sounds good, will refactor.

scheibelp reviewed (changes requested):
I have a couple questions and a couple refactoring requests.

Also, we have CI tests, but IMO it would be good to (temporarily) insert a triggering change into a package definition so that the full CI runs with this diff (it doesn't have to be a low-level package that would lead to many rebuilds).

scheibelp commented on lib/spack/spack/ci.py:
```diff
@@ -1919,32 +1918,49 @@ def reproduce_ci_job(url, work_dir, autostart, gpg_url, runtime):
+                ["echo", f"Re-run install script using:\n\t{install_mechanism}"],
                 # Allow interactive
-                ["exec", "$@"],
+                [exec_cmd, cmd_args, no_exit],
```
(question) this appears to be using `cmd_args` on Windows and Linux: does the definition `cmd_args = "($args -Join ' ')"` apply for both cases?

johnwparent replied:
There's supposed to be an `if IS_WINDOWS else "$@"` after the definition of `cmd_args`, I'll update.

scheibelp commented on lib/spack/spack/ci.py:
```diff
@@ -1973,9 +1989,10 @@ def reproduce_ci_job(url, work_dir, autostart, gpg_url, runtime):
+                entrypoint,
                 job_image,
-                "bash",
+                script,
```
As a general note, IMO the pattern of composing a list from [x, y, z] where each component may or may not be present depending on variables (e.g. platform) can be hard to reason about. In this particular case, I don't see `script` referenced elsewhere, so I think it would be clearer to expand this particular case to `entry_path if IS_WINDOWS else ""`

Also I'm trying to understand the point of having both `--entrypoint` and arguments after the image (this is partially a concern with how things already were, but since we are modifying this code it might be useful to clean it up):

* On Windows, with this PR, you get `--entrypoint powershell.exe <image> script.ps1 powershell.exe`
* On Linux, before this PR, you get `--entrypoint .../entrypoint.sh <image> bash`
  * If `entrypoint.sh` didn't expect any arguments, this probably works fine (the extra "bash" is unnecessary, but would not cause problems)
  * On Linux, with this PR, it looks like you get the same thing

johnwparent replied:
Heard w.r.t. `script`.

W.r.t. `--entrypoint`
The previous (if interactive is true) behavior was to write `exec $@` as the final command of the entrypoint script. `bash` was then passed to it, as `--entrypoint ../entrypoint.sh <image> bash` so the user is dropped into an interactive bash with the setup Spack environment in the container.

The powershell version is the equivalent of this behavior, the only difference is that we need powershell to drive the script since we have no shebangs to leverage, so the script comes after the image where the entrypoint arguments go.

scheibelp reviewed (changes requested):
I have a couple additional questions

johnwparent committed:
define cmd_args for linux

johnwparent committed:
replace script var with value

johnwparent committed:
quote args like before

johnwparent committed:
Merge branch 'develop' of https://github.com/spack/spack into develop

johnwparent commented:
@scheibelp New round of revisions includes your latest review suggestions, dropped cruft commits (zlib churn), and a change to allow Gitlab CI to behave as before on Linux/Mac.
See pipeline status with current set of changes: https://gitlab.staging.spack.io/spack/spack/-/pipelines/952
(there are a few more changes on branch in the pipeline but they're changes that are for GL ci only and don't belong here)

scheibelp commented on lib/spack/spack/ci.py:
```diff
@@ -1919,32 +1919,48 @@ def reproduce_ci_job(url, work_dir, autostart, gpg_url, runtime):
+                ["echo", f"Re-run install script using:\n\t{install_mechanism}"],
                 # Allow interactive
-                ["exec", "$@"],
+                [exec_cmd, cmd_args, no_exit],
```
I have a similar request here as for `process_command`: since each of these 3 elements individually changes based on Windows/not-Windows, can you create this like:

```
if IS_WINDOWS:
  value = [
    "&", "($args -Join ' ')", "-NoExit"
  ]
else:
  value = [...]
```

scheibelp commented on lib/spack/spack/ci.py:
```diff
@@ -1973,9 +1989,10 @@ def reproduce_ci_job(url, work_dir, autostart, gpg_url, runtime):
                 job_image,
-                "bash",
+                entry_path if IS_WINDOWS else "",
+                shell,
```
I think this can be clarified like `process_command` and `entrypoint_script` (my other comment in this review: could you update it like

```
        docker_command = [
            [
                ...
                ":".join([os.path.join(work_dir, "spack"), mount_as_dir, "Z"]),
                "--entrypoint"
            ]
        ]
        if IS_WINDOWS:
            docker_command[0].extend(...)
        else:
            ...
```

scheibelp reviewed (changes requested):
I have a couple refactoring requests related to replacing component-wise Windows/not-Windows customization. Let me know if you disagree.

johnwparent committed:
Refactor entrypoint and docker command assignments

johnwparent committed:
style

scheibelp commented on lib/spack/spack/ci.py:
```diff
@@ -1919,65 +1919,75 @@ def reproduce_ci_job(url, work_dir, autostart, gpg_url, runtime):
+            "--entrypoint",
         ]
+        if IS_WINDOWS:
+            docker_command.extend(["powershell.exe", job_image, entry_script, "powershel.exe"])
```
typo: "powershel.exe"

scheibelp reviewed (changes requested):
There appears to be one typo but once that is fixed this looks great to me

johnwparent committed:
typo

johnwparent commented:
Zlib built, so the changes here clearly function. https://gitlab.staging.spack.io/spack/spack/-/pipelines/960

haampie commented on lib/spack/spack/ci.py:
```diff
@@ -1919,65 +1919,75 @@ def reproduce_ci_job(url, work_dir, autostart, gpg_url, runtime):
+        if IS_WINDOWS:
+            entrypoint_script.extend(["&", "($args -Join ' ')", "-NoExit"])
+        else:
+            entrypoint_script.extend(["exec", "$@"])
```
append not extend.
